### ISSUE TITLE
babel-types typescript generator emits wrong type for Identifier.name

### ISSUE THREAD

**Issue by anilanar** [babel/babel#11100]
TLDR; type generator script falls short when generating a type for `Identifier.name` thus it falls back to `any`.

Proposed change:

1) Define a helper function called refine. I took the name `refine` from [Refinement types](https://en.wikipedia.org/wiki/Refinement_type):

```
export function refine(fn: Validator): Validator {
  function validate(...args) {
    fn(...args);
  }
  validate.refinementOf = fn;
}
```

2) Use `refine` for `Identifier.name` field validator definition:

```
    validate: refine(valueOfType('string'), function(node, key, val) {
        if (!process.env.BABEL_TYPES_8_BREAKING) return;

        if (!esutils.keyword.isIdentifierNameES6(val)) {
          throw new TypeError(`"${val}" is not a valid identifier name`);
        }
      }),
```

3) Update `stringifyValidator`:

```
if (validator.refinementOf) {
  return stringifyValidator(validator.refinementOf, nodePrefix);
}
```

4) (Optional) Fix `chainOf` handling in `stringifyValidator`. It assumes the length of chain is 2 and takes the type info from the second validator. It should take it from the last validator in the chain. After this change, it'd be helpful to take a look at diff of generated types before/after the fix.

5) (Semi-joke) Replace Flow with Typescript.

**Comment by liuxingbaoyu** [comment]
> (Semi-joke) Replace Flow with Typescript.

This has become a reality.😄
